In [1]:
import numpy as np
from decimal import Decimal
from scipy.stats import poisson
from scipy.stats import binom

In [2]:
import scipy.stats as sps

In [3]:
import matplotlib.pyplot as plt

In [4]:
from math import *

In [70]:
def combination (m, n):
    c = np.math.factorial(n)//(np.math.factorial(m)*np.math.factorial(n - m))
    return c

In [6]:
def bernulli (k, n, p):
    P = combination(k, n) * p**k * (1-p)**(n-k)
    return P

In [73]:
def p_poisson(k, p_lambda):
    P = (p_lambda**k) * (np.exp(-p_lambda))/np.math.factorial(k)
    return P

**Задача 1**

Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите `M(X)`, `D(X)`.

Считаем, что выбор правильного ответа случаен для любого из вопросов и вероятность выбора правильного ответа одинакова для всех вопросов и соотетствует 1/4.  
Можно считать, что распределение биноминальное.  
Для данного случая иммет смысл пользоваться формулой Бернулли со следующими параметрами n = 4, p = 1/4 (0.25), k=0...4, где к - количество правильных ответов.



In [8]:
for k in range(5):
    r = bernulli(k, 4, 0.25)
    print(r)

0.31640625
0.421875
0.2109375
0.046875
0.00390625


Таким образом распределение вероятности выбора определенного количества правильных ответов x будет:
<table border="3">
    <tr>
        <td>k</td>
        <td>0</td>
        <td>1</td>
        <td>2</td>
        <td>3</td>
        <td>4</td>
    </tr>
    <tr>
        <td>P(X=k)</td>
        <td>0.31640625</td>
        <td>0.421875</td>
        <td>0.2109375</td>
        <td>0.046875</td>
        <td>0.00390625</td>
    </tr>
</table>

Математическое ожидание для биноминалього распределения:  
M(X) = np

In [9]:
math_expect = 4 * 0.25
print(math_expect)

1.0


Дисперсия для биноминального распределения:

In [10]:
D = 4 * 0.25 * (1-0.25)
print(D)

0.75


**Задача 2**

Вероятность поймать рыбу при одном забрасывании удочки равна 0.01. Какова вероятность того, что число пойманных рыб при 200 забросах удочки будет не менее 5 и не более 10?

Для данного случая подходит распределение Пуассона.  
$λ = np$.  
p = 0.25  
n = 200  
5 <= k <= 10

In [11]:
p_lambda = 200 * 0.01
print(p_lambda)

2.0


Затем складываем вероятности для каждого из событий в интервали от 5 до 10 (вулючительно):

In [12]:
total = 0
for i in range (5, 11):
    probability = p_poisson(i, p_lambda)
    print(probability)
    total += probability

0.03608940886309672
0.012029802954365574
0.0034370865583901638
0.0008592716395975409
0.00019094925324389798
3.8189850648779595e-05


In [13]:
print(f"Общая вероятность выловить от 5 до 10 (включительно) рыб при данных параметрах составляет {round(total, 6) * 100} %")

Общая вероятность выловить от 5 до 10 (включительно) рыб при данных параметрах составляет 5.2645 %


**Задача 3**

В условиях предыдущей задачи сколько забросов удочки нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет поймана рыба?

_Подсказка_. Как посчитать `P(X >= k)` для величины `X`, имеющей биномиальное распределение? Здесь предстоит немножко покодить.


Суммарная вероятность = 0.9, i = 10

P(X=n) = 0.9  
n=?

In [ ]:
p = 0.01
k = 10

С учетом условия задачи рыба должна быть поймана 10 раз или больше, т.е. допустимо решать обратную задачу: т.е. сколько раз нужно закинуть удочку, чтобы поймать меньше 10 рыб.

P(X >= 10) = 1 - P(X < 10)

Суммируем вероятности. Задачу можно попробовать решить методом численного перебора для любого из вариантов распределения.
Пойдем используя тот же метод Пуассона, что для предыдущей задачи.

In [92]:
for n in range (300, 500):
    p_lambda = 0.01 * n
    P = 1 - sum(p_poisson(i, p_lambda) for i in range(10))    
    if P > 0.0099 and P < 0.01:        
        print(f"Полученное значение вероятности {P} достигнуто при {n} попытках")
    if P > 0.01:        
        print(f"Полученное значение вероятности {P} достигнуто при {n} попытках")
        break
        


Полученное значение вероятности 0.009996913843001165 достигнуто при 413 попытках
Полученное значение вероятности 0.010152765154710752 достигнуто при 414 попытках


Таким образом для выполнения условий задачи необходимо порядка 413-414 попыток.